File: androidFood2Fork/src/main/java/com/tfl/food2forkkmm/android/presentation/recipe_list/RecipeListViewModel.kt
package com.tfl.food2forkkmm.android.presentation.recipe_list

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tfl.food2forkkmm.domain.model.Recipe
import com.tfl.food2forkkmm.interactors.recipe_list.SearchRecipes
import com.tfl.food2forkkmm.presentation.recipe_list.FoodCategory
import com.tfl.food2forkkmm.presentation.recipe_list.RecipeListEvents
import com.tfl.food2forkkmm.presentation.recipe_list.RecipeListState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import javax.inject.Inject

@HiltViewModel
class RecipeListViewModel
@Inject
constructor(private val savedStateHandle: SavedStateHandle,
            private val searchRecipes: SearchRecipes): ViewModel() {

    val state: MutableState<RecipeListState> = mutableStateOf(RecipeListState())

    init {
        onTriggerEvents(RecipeListEvents.LoadRecipes)
    }

    fun onTriggerEvents(event: RecipeListEvents) {
        when(event) {
            RecipeListEvents.LoadRecipes -> {
                loadRecipes()
            }
            RecipeListEvents.NextPage -> {
                nextPage()
            }
            RecipeListEvents.NewSearch -> {
                nextSearch()
            }
            is RecipeListEvents.OnUpdateQuery -> {
                state.value = state.value.copy(query = event.query, selectedCategory = null)
            }
            is RecipeListEvents.OnSelectCategory -> {
                onSelectCategory(event.category)
            }
            else -> {
                handleError("Invalid Event")
            }
        }
    }

    private fun onSelectCategory(category: FoodCategory) {
        state.value = state.value.copy(selectedCategory = category, query = category.value)
        nextSearch()
    }

    private fun nextSearch() {
        state.value = state.value.copy(page = 1, recipes = listOf())
        loadRecipes()
    }

    private fun handleError(errorMessage: String) {
        println(errorMessage)
    }

    private fun nextPage() {
        state.value = state.value.copy(page = state.value.page + 1)
        loadRecipes()
    }

    private fun loadRecipes() {
        searchRecipes.execute(page = state.value.page, query = state.value.query)
            .onEach { dataState ->
                state.value = state.value.copy(isLoading = dataState.isLoading)
                dataState.data?.let { recipes ->
                    appendRecipes(recipes = recipes)
                }
                dataState.message.let { message ->
                    handleError("RecipeListVM: ${message}")
                }
            }
            .launchIn(viewModelScope)
    }

    private fun appendRecipes(recipes: List<Recipe>) {
        val curr = ArrayList(state.value.recipes)
        curr.addAll(recipes)
        state.value = state.value.copy(recipes = curr)
    }
}

File: iosFood2Fork/iosFood2Fork/presentation/recipe_list/RecipeListViewModel.swift
//
//  RecipeListViewModel.swift
//  iosFood2Fork
//
//  Created by Viktor Yamchinov on 08.04.2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared

class RecipeListViewModel: ObservableObject {
    
    let searchRecipes: SearchRecipes
    let foodCategoryUtil: FoodCategoryUtil
    
    @Published var state: RecipeListState = RecipeListState()
    
    init(searchRecipes: SearchRecipes, foodCategoryUtil: FoodCategoryUtil) {
        self.searchRecipes = searchRecipes
        self.foodCategoryUtil = foodCategoryUtil
        onTriggerEvent(stateEvent: RecipeListEvents.LoadRecipes())
    }
    
    func onTriggerEvent(stateEvent: RecipeListEvents) {
        switch stateEvent {
        case is RecipeListEvents.LoadRecipes:
            loadRecipes()
        case is RecipeListEvents.NewSearch:
            doNothing()
        case is RecipeListEvents.NextPage:
            doNothing()
        case is RecipeListEvents.OnUpdateQuery:
            doNothing()
        case is RecipeListEvents.OnSelectCategory:
            doNothing()
        case is RecipeListEvents.OnRemoveHeadMessageFromQueue:
            doNothing()
        default:
            doNothing()
        }
    }
    
    private func loadRecipes() {
        let currentState = self.state.copy() as! RecipeListState
        do {
            try searchRecipes.execute(page: Int32(currentState.page), query: currentState.query)
                .collectCommon(coroutineScope: nil, callback: {dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        let loading = dataState?.isLoading ?? false
                        self.updateState(isLoading: loading)
                        if data != nil {
                            self.appendRecipes(recipes: data as! [Recipe])
                        }
                        if message != nil {
                            self.handleMessageByUiComponentType(message!.build())
                        }
                    }
                })
        } catch {
            print("\(error)")
        }
    }
    
    func doNothing() {}
    
    func appendRecipes(recipes: [Recipe]) {
        var currentState = self.state.copy() as! RecipeListState
        var currentRecipes = currentState.recipes
        currentRecipes.append(contentsOf: recipes)
        self.state = self.state.doCopy(isLoading: currentState.isLoading, page: currentState.page, query: currentState.query, selectedCategory: currentState.selectedCategory, recipes: currentRecipes, queue: currentState.queue)
    }
    
    private func handleMessageByUiComponentType(_ message: GenericMessageInfo) {
        
    }
    
    func updateState(isLoading: Bool? = nil, page: Int? = nil, query: String? = nil, queue: Queue<GenericMessageInfo>? = nil) {
        let currentState = self.state.copy() as! RecipeListState
        self.state = self.state.doCopy(isLoading: isLoading ?? currentState.isLoading, page: Int32(page ?? Int(currentState.page)), query: query ?? currentState.query, selectedCategory: currentState.selectedCategory, recipes: currentState.recipes, queue: queue ?? currentState.queue)
    }
}
